samples/shared/DefaultAudioStreamCallback.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SAMPLES_DEFAULT_AUDIO_STREAM_CALLBACK_H
+#define SAMPLES_DEFAULT_AUDIO_STREAM_CALLBACK_H
+
+
+#include <oboe/AudioStreamCallback.h>
+#include <shared/AudioEngineBase.h>
+#include <shared/IRenderableAudio.h>
+#include "../debug-utils/logging_macros.h"
+
+class DefaultAudioStreamCallback : public oboe::AudioStreamCallback {
+public:
+    DefaultAudioStreamCallback(AudioEngineBase &audioEngineBase):
+        mParent(audioEngineBase) { }
+
+    virtual oboe::DataCallbackResult
+    onAudioReady(oboe::AudioStream *oboeStream, void *audioData, int32_t numFrames) override {
+        float *outputBuffer = static_cast<float *>(audioData);
+        if (!mRenderable){
+            return oboe::DataCallbackResult ::Stop;
+        }
+        mRenderable->renderAudio(outputBuffer, numFrames);
+        return oboe::DataCallbackResult::Continue;
+    }
+    virtual void onErrorAfterClose(oboe::AudioStream *oboeStream, oboe::Result error) override {
+        // Restart the stream when it errors out with disconnect
+        if (error == oboe::Result::ErrorDisconnected) {
+            LOGE("Restarting AudioStream after disconnect");
+            mParent.restartStream();
+        }
+    }
+
+    void setSource(IRenderableAudio *renderable) { mRenderable = renderable;}
+    virtual void onSetupComplete() {}
```

Comment:
Is this method ever used?

Reply:
Used in engine construction.

Reply:
I know it's called in `createPlaybackStream` but no code is executed because the method does nothing. What's the point of it?

Reply:
It is for subclasses to override. Any final setup construction. In the base class does nothing.

Reply:
I would remove it unless it's actually doing something. We can add it back in if/when there's a need for it.